### DISCUSSION TITLE
Module Not Found Error, while running workflows in FlyteConsole

### DISCUSSION THREAD

**Discussion by sakthi-tech4th** [flyteorg/flyte#2909] [Q&A]
Hi,

I created a simple worflow with one task, that reads a csv file into a pandas dataframe.
My code is something like below.
`
import os
import sys
import pandas as pd
sys.path.append(r'path_to_my_python_modules')

from data.data_ingestion import load_data # data is folder in path_to_my_models, which contains the file data_ingestion.py. 

@task
def getdata(data_path: str='', some_int: int=0) -> pd.DataFrame:
      df =load_data(data_path=data_path, some_int=some_int)
      return df

@workflow
def my_wf(my_path: str='', my_int: int=0) -> pd.DataFrame:
      df = getdata(data_path=my_path, some_int=my_int)
      return df

if __name__ == "__main__":
    data_root = os.getenv('raw_data_folder')
    data_path = os.path.join(data_root, 'my_data.csv')
    print(data_path)
    some_int= 6
    data_df = my_wf(my_path=data_path, my_int=some_int)
    print(data_df)
`
When i run the code locally using:
'python my_pipeline.py'
I have no errors, outputs are as expected.

However, when I run pyflyte:
`pyflyte run --remote my_pipeline.py my_wf`

I get Module Not Found Error.
    from data.data_ingestion import load_data
ModuleNotFoundError: No module named 'data'

I believe, I have include my modules in the dockerimage/container where my code is running.
Could you please guide me on how to make my modules available to the workflow?

**Comment by samhita-alla** [comment]
@sakthi-tech4th, `pyflyte run` command operates on a single file. For you to be able to use code in a different module, you'll have to run `pyflyte register` command: `pyflyte register <package-name>`.

**Reply by laruokol** [reply]
@samhita-alla 
I'm trying to do a similar thing. I have the following structure for my code:

```
project_root/
├── flyte/
   ├── __init__.py
   ├── tasks.py
   ├── utils.py
   └── workflow.py
```
I have run the `./docker_build_and_tag.sh` script to build a container image. 
Then packaged the workflow:
`pyflyte --pkgs flyte package --image <image-name>`
registered the workflow:
`flytectl register files --project mlexample --domain development --archive flyte-package.tgz --version v1`
registered the code (as instructed above):
`pyflyte register flyte --project mlexample --domain development`
 created a launchplan:
`flytectl get launchplan --project mlexample --domain development src.workflow.wf --latest --execFile exec_spec.yaml`
and created an execution:
`flytectl create execution --project mlexample --domain development --execFile exec_spec.yaml`

I have also tried to `pyflyte run --remote Flyte/workflow.py wf`, but all attempts end with `ModuleNotFoundError`, when tasks are imported from `flyte.tasks` in `workflow.py`. I'm pretty much out of ideas here.

**Comment by samhita-alla** [comment]
@laruokol, you're trying to use code from a different Python module. `pyflyte register` is the preferred command for this case. `pyflyte run` is for single files. `pyflyte package` + `flytectl register` works as well but `pyflyte register` is the easiest to get started with and test. So let's zero in on the `pyflyte register` command. 

How are you importing tasks in the `workflow.py` file? Have you tried doing a relative import: `from .tasks import ...`?

**Reply by laruokol** [reply]
Thanks for your reply @samhita-alla 👍 
I have in the main `workflow.py` import defined as:
```
try:
    from .tasks import (
        ingest_data,
        preprocess,
        split_data,
        train_model,
        evaluate_model,
    )
except ImportError:
    from tasks import (
        ingest_data,
        preprocess,
        split_data,
        train_model,
        evaluate_model,
    )
```
I have tried to understand this better by using a much simpler example, with all the code written in the same file, which I have registered as `pyflyte register flyte/simple.py`  However, even this fails to run on Flyte sandbox:
```
currentAttempt done. Last Error: USER::Pod failed. No message received from kubernetes.
[a66jnhxk6fpbcnr98lqz-n0-0] terminated with exit code (1). Reason [Error]. Message: 
unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/root/flyte/simple.py", line 5, in <module>
    from sklearn.model_selection import train_test_split
ModuleNotFoundError: No module named 'sklearn'
```
I have tried to fix this by providing a container image that I know contains `sklearn` with any success.

**Reply by samhita-alla** [reply]
If you're providing a container image that has sklearn in it, this issue shouldn't crop up. Have you tried using `ghcr.io/flyteorg/flytecookbook:core-latest` as your image? The command would then be `pyflyte register --image ghcr.io/flyteorg/flytecookbook:core-latest flyte/simple.py`.

**Reply by laruokol** [reply]
Yep, I have tried that container, with the same result: sklearn was not found :(

**Reply by laruokol** [reply]
My `simple.py` looks like this:
```
from flytekit import workflow, task
from typing import Tuple
import pandas as pd
from sklearn.model_selection import train_test_split


@task
def get_data(uri: str) -> pd.DataFrame:
    data = pd.read_csv(uri)
    return data


@task
def summarize_data(df: pd.DataFrame) -> pd.DataFrame:
    return df.describe()


@task
def split_data(
    df: pd.DataFrame, test_size: float = 0.2
) -> Tuple[pd.DataFrame, pd.DataFrame]:
    return train_test_split(
        df,
        test_size=test_size,
    )  # type: ignore


@workflow
def wf(
    uri: str = "https://gist.githubusercontent.com/curran/a08a1080b88344b0c8a7/raw/0e7a9b0a5d22642a06d3d5b9bcbad9890c8ee534/iris.csv",
    test_size: float = 0.2,
) -> pd.DataFrame:
    data = get_data(uri=uri)
    splits = split_data(df=data, test_size=test_size)
    summary = summarize_data(df=splits[0])
    return summary  # type: ignore


if __name__ == "__main__":
    out = wf()  # type: ignore
    print(out)
```

**Reply by samhita-alla** [reply]
@laruokol, it works for me. This is what I did:

* `flytectl demo start`
* `pyflyte register --image ghcr.io/flyteorg/flytecookbook:core-latest core/flyte_basics/test.py`

<img width="1444" alt="image" src="https://user-images.githubusercontent.com/27777173/193544792-08d82389-195a-4174-a7f1-45f4fb955198.png">

**Reply by laruokol** [reply]
Hi @samhita-alla,
it is reassuring that this works for you. For me, however, the first task is still running after 13 minutes! 
Could it be that that there is an incompatibility issue with my machine, running Apple M2 chip?

**Reply by samhita-alla** [reply]
Um, could be the case. Could you inspect the pod logs? 

1. `docker ps`
2. Get the container ID
3. `docker exec -it <container-id> bash`
4. `kubectl get pods -n flytesnacks-development`
5. Get the concerning pod name
6. `kubectl logs <pod-name> -n flytesnacks-development`

**Reply by laruokol** [reply]
Okey,
I spanned up the cluster as `flytectl demo start` and registered the code. Then I launched an execution and went see what is going on in the sandbox-container. The error given by
`kubectl logs a76x7xb97cggjzhhmntj-n0-0 -n flytesnacks-development`
was something different that I have encountered before:
```tar: Removing leading `/' from member names```

Next I restarted the cluster with `flytectl sandbox start --source .`. This time I get
```
kubectl get pods -n flytesnacks-development
NAME                        READY   STATUS    RESTARTS   AGE
ahsv2fg5q2mdf5glmm8q-n0-0   1/1     Running   0          11m
```
but obviously nothing is happening (the status comment given in the UI is: [ContainersNotReady|ContainerCreating]: containers with unready status: [ahsv2fg5q2mdf5glmm8q-n0-0]|). 
Running `kubectl logs ahsv2fg5q2mdf5glmm8q-n0-0 -n flytesnacks-development` returns nothing.

**Reply by samhita-alla** [reply]
Um, you can ignore the tar warning. So the issue now is that the pod's taking forever to get started. Let me ask my team to take a look at this. It seems like an issue with M2 chip cause the code works fine on intel.

**Reply by samhita-alla** [reply]
@laruokol, could you wait some time for it to get started -- 30 mins or so? Let's see if it works then.

**Reply by laruokol** [reply]
However, 
I decided to give custom packaging a one more try: 
```
pyflyte --pkgs flyte.workflows package --image "my_flyte_project:v1"
pyflyte register --image "my_flyte_project:v1" src/test.py
```
This did not work, but, from the error I deduced that the python version that the container was built on was too old. Setting that to a newer version (3.10) did the trick!

**Reply by laruokol** [reply]
Thank you @samhita-alla for your support on this issue :)

**Reply by samhita-alla** [reply]
Awesome! Glad that you got it working.

**Comment by highfly22** [comment]
Just add ENV PYTHONPATH=/root in your Dockerfile.
